File: tools/bigquery-zos-mainframe-connector/src/main/scala/com/google/cloud/bqsh/BQ.scala
/*
 * Copyright 2019 Google LLC All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.cloud.bqsh

import com.google.api.gax.retrying.RetrySettings
import com.google.api.gax.rpc.FixedHeaderProvider
import com.google.auth.Credentials
import com.google.cloud.RetryOption
import com.google.cloud.bigquery.{BigQuery,BigQueryError,BigQueryException,BigQueryOptions,DatasetId,Field,FieldList,JobInfo,Job,JobId,JobStatus,QueryJobConfiguration,Schema,StandardSQLTypeName,TableId}
import com.google.cloud.gszutil.CCATransportFactory
import com.google.cloud.http.HttpTransportOptions
import com.google.common.base.Preconditions
import com.google.common.collect.ImmutableList
import org.threeten.bp.Duration
import scala.collection.JavaConverters._

object BQ {
  def defaultClient(project: String, location: String, credentials: Credentials): BigQuery = {
    BigQueryOptions.newBuilder
      .setLocation(location)
      .setProjectId(project)
      .setCredentials(credentials)
      .setTransportOptions(HttpTransportOptions.newBuilder
        .setHttpTransportFactory(new CCATransportFactory)
        .build)
      .setRetrySettings(RetrySettings.newBuilder
        .setMaxAttempts(8)
        .setTotalTimeout(Duration.ofMinutes(30))
        .setInitialRetryDelay(Duration.ofSeconds(8))
        .setMaxRetryDelay(Duration.ofSeconds(32))
        .setRetryDelayMultiplier(2.0d)
        .build)
      .setHeaderProvider(FixedHeaderProvider.create("user-agent", Bqsh.UserAgent))
      .build
      .getService
  }

  def runJob(bq: BigQuery, cfg: QueryJobConfiguration, jobId: JobId, timeoutSeconds: Long, sync: Boolean): Job = {
    require(bq != null, "BigQuery must not be null")
    require(cfg != null, "QueryJobConfiguration must not be null")
    require(jobId != null, "JobId must not be null")
    try {
      val job = bq.create(JobInfo.of(jobId, cfg))
      if (sync) await(job, jobId, timeoutSeconds)
      else job
    } catch {
      case e: BigQueryException =>
        if (e.getReason == "duplicate" && e.getMessage.startsWith("Already Exists: Job")) {
          await(bq, jobId, timeoutSeconds)
        } else {
          throw new RuntimeException(e)
        }
    }
  }

  def await(bq: BigQuery, jobId: JobId, timeoutSeconds: Long): Job = {
    Preconditions.checkNotNull(jobId)
    val job = bq.getJob(jobId)
    await(job, jobId, timeoutSeconds)
  }

  def await(job: Job, jobId: JobId, timeoutSeconds: Long): Job = {
    Preconditions.checkNotNull(jobId)
    if (job == null) {
      throw new RuntimeException(s"Job ${jobId.getJob} doesn't exist")
    } else {
      job.waitFor(RetryOption.totalTimeout(Duration.ofSeconds(timeoutSeconds)))
    }
  }

  def isDone(job: Job): Boolean = {
    job.getStatus.getState == JobStatus.State.DONE
  }

  def hasError(job: Job): Boolean = {
    job.getStatus.getError != null || hasExecutionErrors(job)
  }

  def hasExecutionErrors(job: Job): Boolean = {
    val status = job.getStatus
    val errors = status.getExecutionErrors
    errors != null && errors.size() > 0
  }

  case class BQError(message: scala.Option[String],
                     reason: scala.Option[String],
                     location: scala.Option[String]) {
    override def toString: String = {
      s"BigQueryError:\nreason: ${reason.getOrElse("no reason given")}\nmessage: ${message.getOrElse("no message given")}"
    }
  }

  case class BQStatus(state: JobStatus.State,
                      error: scala.Option[BQError],
                      executionErrors: Seq[BQError]) {
    def hasError: Boolean = error.isDefined || executionErrors.nonEmpty
  }

  def getExecutionErrors(status: JobStatus): Seq[BQError] = {
    val e = status.getExecutionErrors
    if (e == null) Seq.empty
    else e.asScala.flatMap(toBQError)
  }

  def toBQError(error: BigQueryError): scala.Option[BQError] = {
    if (error == null) None
    else scala.Option(BQError(
      message = scala.Option(error.getMessage),
      reason = scala.Option(error.getReason),
      location = scala.Option(error.getLocation)
    ))
  }

  def getStatus(job: Job): scala.Option[BQStatus] = {
    val status = job.getStatus
    if (status == null) None
    else {
      scala.Option(BQStatus(
        state = status.getState,
        error = toBQError(status.getError),
        executionErrors = getExecutionErrors(status)))
    }
  }

  def throwOnError(status: BQStatus): Unit = {
    val sb = new StringBuilder
    if (status.hasError) {
      status.error.foreach { err =>
        sb.append(err.toString)
      }
      if (status.executionErrors.nonEmpty) {
        sb.append("Execution Errors:\n")
      }
      status.executionErrors.foreach { err =>
        sb.append(err.toString)
      }
      throw new RuntimeException(sb.result())
    }
  }

  def throwOnError(job: Job): Unit = {
    getStatus(job) match {
      case Some(status) =>
        throwOnError(status)
      case _ =>
        throw new RuntimeException("missing job status")
    }
  }

  /**
    *
    * @param datasetSpec [PROJECT]:[DATASET] may override project
    * @param defaultProject BigQuery billing project
    * @return
    */
  def resolveDataset(datasetSpec: String, defaultProject: String): DatasetId = {
    val i = datasetSpec.indexOf(':')
    val project = if (i > 0) datasetSpec.substring(0, i) else defaultProject
    val dataset = datasetSpec.substring(i+1)
    DatasetId.of(project, dataset)
  }

  /**
    *
    * @param tableSpec [PROJECT]:[DATASET].TABLE may override dataset and project
    * @param defaultProject BigQuery billing project
    * @param defaultDataset [PROJECT]:[DATASET] may override project
    * @return
    */
  def resolveTableSpec(tableSpec: String, defaultProject: String, defaultDataset: String): TableId = {
    val i = tableSpec.indexOf(':')
    val j = tableSpec.indexOf('.')

    val default = resolveDataset(defaultDataset, defaultProject)

    val project =
      if (i > 0) tableSpec.substring(0, i)
      else default.getProject

    val dataset =
      if (j > 0) tableSpec.substring(i+1, j)
      else default.getDataset

    val table =
      if (j > 0) tableSpec.substring(j+1, tableSpec.length)
      else tableSpec
    TableId.of(project, dataset, table)
  }

  def parseSchemaUpdateOption(schemaUpdateOption: Seq[String]): java.util.List[JobInfo.SchemaUpdateOption] = {
    val schemaUpdateOptionsBuilder = ImmutableList.builder[JobInfo.SchemaUpdateOption]()
    if (schemaUpdateOption.contains(JobInfo.SchemaUpdateOption.ALLOW_FIELD_ADDITION.name())) {
      schemaUpdateOptionsBuilder
        .add(JobInfo.SchemaUpdateOption.ALLOW_FIELD_ADDITION)
    }
    if (schemaUpdateOption.contains(JobInfo.SchemaUpdateOption.ALLOW_FIELD_RELAXATION.name())) {
      schemaUpdateOptionsBuilder
        .add(JobInfo.SchemaUpdateOption.ALLOW_FIELD_RELAXATION)
    }
    schemaUpdateOptionsBuilder.build
  }

  def parseField(s: String): Field = {
    val Array(field,dataType) = s.split(':')
    val fieldList = FieldList.of()
    val typeName = StandardSQLTypeName.valueOf(dataType)
    Field.newBuilder(field,typeName,fieldList).build
  }

  def parseSchema(s: Seq[String]): Schema = {
    Schema.of(s.map(parseField):_*)
  }

  def isValidTableName(s: String): Boolean = {
    s.matches("[a-zA-Z0-9_]{1,1024}")
  }

  def isValidBigQueryName(s: String): Boolean =
    s.matches("[a-zA-Z0-9_]{1,1024}")
}

File: tools/data-modeling-agent/src/agents/data_modelling_agent/sub_agents/ddl_agent/utils/bq.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from google.cloud import bigquery


def cleanup_ddl(ddl, project_id, dataset_id):
    ddl_lines = ddl.split("\n")
    cleaned_ddl_lines = []
    for line in ddl_lines:
        if "```"in line:
            continue
        if "CREATE TABLE" in line:
            if "project_id" in line:
                line = line.replace("project_id", project_id)
            if "dataset_id" in line:
                line = line.replace("dataset_id", dataset_id)
        cleaned_ddl_lines.append(line.strip())
    cleaned_ddl = "\n".join(cleaned_ddl_lines)
    return cleaned_ddl
    
def validate_ddl(ddl_statement):
    client = bigquery.Client()
    try:
        job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
        query_job = client.query(ddl_statement, job_config=job_config)
        query_job.result()  # This will raise an exception if the DDL is invalid
        print("DDL statement is valid.")
    except Exception as e:
        print(f"DDL validation failed: {e}")
        return False
    return True


def create_bigquery_dataset(project_id, dataset_id, dataset_to_be_deleted, location="US"):
    """Creates a BigQuery dataset in the specified project and location."""
    client = bigquery.Client(project=project_id)

    # Construct a full Dataset object to specify location
    dataset_ref = bigquery.DatasetReference(project_id, dataset_id)
    dataset = bigquery.Dataset(dataset_ref)
    dataset.location = location
    if dataset_to_be_deleted:
        try:
            client.delete_dataset(dataset, delete_contents=True, not_found_ok=True)
        except Exception as e:
            print(f"Error deleting dataset: {e}")
            return {
                'status': 'error',
                "message": f"Dataset {dataset_id} could not be deleted in project {project_id}.",
            }
        print(f"Deleted dataset {dataset_id} in project {project_id}.")
    try:
        dataset = client.create_dataset(dataset)  # API request
        print(f"Dataset {dataset.dataset_id} created in project {client.project} at location {dataset.location}.")
    except Exception as e:
        print(f"Error creating dataset: {e}")
        if "Already Exists" in str(e):
            print(f"Dataset {dataset_id} already exists in project {project_id}.")
            return {
                'status': 'error',
                "message": f"Dataset {dataset_id} already exists in project {project_id}. Do you want to delete the dataset and create a new one?",
            }
    return None

def execute_bq_ddl(ddl_statement):
        # Initialize a BigQuery client
    client = bigquery.Client()
    try:
    # Execute the DDL query
    # The .result() method blocks until the job completes, providing job status and results.
        query_job = client.query(ddl_statement)
        query_job.result()  # Wait for the job to complete
    except Exception as e:
        print(f"DDL statement execution failed: {e}")
        return False

    print("DDL statement executed successfully.")
    return True